pkg/pool-weighted/contracts/smart/ManagedPool.sol

```diff
@@ -265,7 +326,9 @@ contract ManagedPool is BaseWeightedPool, ProtocolFeeCache, ReentrancyGuard {
     }
 
     function _getTotalTokens() internal view virtual override returns (uint256) {
-        return _getMiscData().decodeUint7(_TOTAL_TOKENS_OFFSET);
+        (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
```

Comment:
My first instinct would be to cache this in storage on the pool contract. We're going to be performing a lot of SLOADs to get the number of tokens here.